confdgnmi/src/confd_gnmi_api_adapter.py

```diff
@@ -420,95 +422,93 @@ def capabilities(self):
-                parent_path_cand == "" and elem.tag == "{http://tail-f.com/ns/config/1.0}config"):
-            parent_path_cand += "/" + elem.tag.split("}")[-1]
-        if len(elem) > 0:
-            # we (simplification) assume that if there are more sibling elements,
-            # this is list and first element is key, this should be fixed with
-            # help of `cs_nodes`
-            # in this way we can detect some (non list) paths incorrectly, we try to fix
-            # by comparing original path
-            if len(elem) > 1:
-                if elem[0].text is not None:
-                    parent_path_cand2 = parent_path_cand + "[{}={}]".format(
-                        elem[0].tag.split("}")[-1], elem[0].text)
-                    if parent_path_cand2.startswith(
-                            orig_path_str) or \
-                            orig_path_str.startswith(parent_path_cand2):
-                        parent_path_cand = parent_path_cand2
-
-            for e in elem:
-                path_vals.extend(self.get_xml_paths(e, orig_path_str,
-                                                    parent_path_cand))
-        else:
-            # check if this is path for exact key
-            keypath_cand = None
-            if elem.text is not None:
-                val = elem.text.split(":")[-1]
-                path_list = parent_path_cand.split('/')
-                if len(path_list) >= 2:
-                    keypath_cand = "/".join(path_list[:-1]) + "[{}={}]".format(
-                        path_list[-1], val) + "/" + path_list[-1]
-                if keypath_cand is not None and orig_path_str == keypath_cand:
-                    path = keypath_cand
-                    parent_path_cand = path
-                else:
-                    path = parent_path_cand
-                # if path != orig_path_str:
-                #     path += "/" + elem.tag.split("}")[-1]
-                # we also need to check possibility this is exact key leaf
-                # skip not correctly detected paths
-                if path.startswith(parent_path_cand):
-                    path_vals.append((path, val))
-        log.debug("<== paths=%s", path_vals)
-        return path_vals
-
-    def get_updates_with_maapi_save(self, path, prefix, data_type):
-        log.debug("==> path=%s prefix=%s data_type=%s", path, prefix, data_type)
-        path_str = make_xpath_path(path, prefix, quote_val=False)
-        # we need ariant with quoted values for maapi_save
-        path_str_quote = make_xpath_path(path, prefix, quote_val=True)
-        prefix_str = make_xpath_path(gnmi_prefix=prefix)
-        log.debug("path_str=%s", path_str)
-        save_flags = _confd.maapi.CONFIG_XML | _confd.maapi.CONFIG_XPATH
+        container_flags = (_confd.CS_NODE_IS_LIST | _confd.CS_NODE_IS_CONTAINER)
+        is_container = csnode.info().flags() & container_flags != 0
+        add_update = add_update_json if is_container else add_update_value
```

Comment:
This should be given by requested encoding, not by node type.

Reply:
fixed - everything is json_ietf now